File: swift/2021/Day01-2021.swift
import Foundation

func Day01_2021() {
    let input = try! String(contentsOfFile: "2021/day01.data")
    let numbers = input.split(separator: "\n").compactMap { Int($0) }
    
    Part1(numbers: numbers)
    Part2(numbers: numbers)
}

fileprivate func Part1(numbers: [Int]) {
    var numberOfIncreases = 0
    for i in 1 ..< numbers.count {
        if numbers[i] > numbers[i - 1] {
            numberOfIncreases += 1
        }
    }
    
    print("\(numberOfIncreases)")
}

fileprivate func Part2(numbers: [Int]) {
    let windowSize = 3
    
    var numberOfIncreases = 0
    
    var previous = numbers[0 ..< windowSize].reduce(0, +)
    
    for i in 1 ... numbers.count - windowSize {
    
        let current = numbers[i ..< i + windowSize].reduce(0, +)
        
        if current > previous {
            numberOfIncreases += 1
        }
        
        previous = current
    }
    
    print("\(numberOfIncreases)")
}

File: rust/2021/Day01-2021.rs
use std::path::Path;
use std::io::prelude::*;
use std::fs::File;

fn main() {
    
    let path = Path::new("../../data/2021/day01.data");
    let mut file = match File::open(&path) {
        Err(why) => panic!("couldn't open {}: {}", path.display(), why),
        Ok(file) => file,
    };

    let mut s = String::new();
    match file.read_to_string(&mut s) {
        Err(why) => panic!("Couldn't read {}: {}", path.display(), why), 
        Ok(_) => (), 
    }

    let mut numbers = Vec::new();
    for number in s.lines() {
        numbers.push(number.parse::<u32>().unwrap())
    }

    part1(&numbers);
    part2(&numbers);
}

fn part1(numbers: &Vec<u32>) {
    let mut total = 0;

    for i in 1 .. numbers.len() {
        if numbers[i] > numbers[i - 1] {
            total += 1;
        }
    }

    println!("{}", total);
}

fn part2(numbers: &Vec<u32>) {
    let window_size = 3;

    let mut total = 0;

    let mut previous = 0;
    for i in 0 .. window_size {
        previous += numbers[i];
    }

    for i in 1 ..= numbers.len() - window_size {

        let mut current = 0;
        for j in i .. i + window_size {
            current += numbers[j];
        }

        if current > previous {
            total += 1;
        }

        previous = current;

    }

    println!("{}", total);
}